src/core/meshcop/dataset_manager.cpp

```diff
@@ -531,51 +531,64 @@ void DatasetManager::HandleMgmtSetResponse(Coap::Message *aMessage, const Ip6::M
 
-    SuccessOrExit(Tlv::FindTlvValueOffsetRange(aMessage, Tlv::kGet, offsetRange));
+    VerifyOrExit(response != nullptr);
+    SuccessOrExit(error = Get<Tmf::Agent>().SendMessage(*response, aMessageInfo));
 
-    while (!offsetRange.IsEmpty())
-    {
-        IgnoreError(aMessage.Read(offsetRange, tlvType));
-        tlvList.Add(tlvType);
-        offsetRange.AdvanceOffset(sizeof(uint8_t));
-    }
-
-    // MGMT_PENDING_GET.rsp must include Delay Timer TLV (Thread 1.1.1
-    // Section 8.7.5.4).
-
-    if (!tlvList.IsEmpty() && IsPendingDataset())
-    {
-        tlvList.Add(Tlv::kDelayTimer);
-    }
+    LogInfo("sent %s dataset get response to %s", IsActiveDataset() ? "active" : "pending",
+            aMessageInfo.GetPeerAddr().ToString().AsCString());
 
 exit:
-    SendGetResponse(aMessage, aMessageInfo, tlvList);
+    FreeMessageOnError(response, error);
 }
 
-void DatasetManager::SendGetResponse(const Coap::Message    &aRequest,
-                                     const Ip6::MessageInfo &aMessageInfo,
-                                     const TlvList          &aTlvList) const
+Coap::Message *DatasetManager::ProcessGetRequest(const Coap::Message &aRequest) const
 {
-    Error          error = kErrorNone;
-    Coap::Message *message;
+    // Processes a MGMT_ACTIVE_GET or MGMPT_PENDING_GET request
+    // and prepares the response.
+
+    Error          error    = kErrorNone;
+    Coap::Message *response = nullptr;
     Dataset        dataset;
+    TlvList        tlvList;
+    OffsetRange    offsetRange;
 
-    IgnoreError(Read(dataset));
+    if (Tlv::FindTlvValueOffsetRange(aRequest, Tlv::kGet, offsetRange) == kErrorNone)
+    {
+        while (!offsetRange.IsEmpty())
+        {
+            uint8_t tlvType;
 
-    message = Get<Tmf::Agent>().NewPriorityResponseMessage(aRequest);
-    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+            IgnoreError(aRequest.Read(offsetRange, tlvType));
+            tlvList.Add(tlvType);
+            offsetRange.AdvanceOffset(sizeof(uint8_t));
+        }
+
+        // MGMT_PENDING_GET.rsp must include Delay Timer TLV (Thread 1.1.1
+        // Section 8.7.5.4).
+
+        if (!tlvList.IsEmpty() && IsPendingDataset())
+        {
+            tlvList.Add(Tlv::kDelayTimer);
+        }
+    }
+
+    // Prepare the response
+
+    SuccessOrExit(error = Read(dataset));
+
+    response = Get<Tmf::Agent>().NewPriorityResponseMessage(aRequest);
+    VerifyOrExit(response != nullptr, error = kErrorNoBufs);
 
     for (const Tlv *tlv = dataset.GetTlvsStart(); tlv < dataset.GetTlvsEnd(); tlv = tlv->GetNext())
     {
         bool shouldAppend = true;
 
-        if (!aTlvList.IsEmpty())
+        if (!tlvList.IsEmpty())
         {
-            shouldAppend = aTlvList.Contains(tlv->GetType());
+            shouldAppend = tlvList.Contains(tlv->GetType());
         }
 
         if ((tlv->GetType() == Tlv::kNetworkKey) && !Get<KeyManager>().GetSecurityPolicy().mObtainNetworkKeyEnabled)
```

Comment:
when request dataset from BA locally, `mObtainNetworkKeyEnabled` should be ignored and networkkey should be appended in the response

Reply:
Good point. However, I suggest keeping this the same for now in this PR.

The existing behavior in `main`, where the border agent would forward the request to the leader, would already apply this flag since the leader would see the request originating from the BA.

We need to consider how this can be checked when a message is forwarded by the BA. This would be worth addressing in its own PR.

Additionally, I'd like us to clarify the intention behind this flag and whether it makes sense to update this in the specification. It doesn't seem logical to me to apply this only to requests from other Thread nodes in the network (which supposedly already know the full Dataset) but allow an external commissioner to retrieve the key when this flag is set.

Reply:
Not fully get the last two paragraphs.

Ignoring `mObtainNetworkKeyEnabled` seems closely related to the process of MGMT_*_GET locally on BA as in SPEC-1216, 8.4.9.4.1.4
```
when processing the TMF requests MGMT_ACTIVE_GET.req and MGMT_PENDING_GET.req,
the request is processed as if the value of the Security Policy O-bit is assumed to be
alwayswere '1' (OOB enabled) so that the Network Key can be provided to the Candidate.
The actual value of the O-bit in the Security Policy TLV is ignored for this by the BA.```

And when MGMT_*GET is processed on leader, this bit should be respected as previously.

Reply:
Yes, thanks. Now I recall the discussion in SPEC-1216 (regarding how to handle the `O` flag). 

I misremembered this, assuming we needed to apply the same logic when the request is forwarded to the leader.